version: '3.8'

services:
  db:
    image: postgres:14
    container_name: dapp-for-business_db_1
    networks:
      - dapp-for-business_dapp-network
    environment:
      POSTGRES_USER: user
      POSTGRES_PASSWORD: password
      POSTGRES_DB: dappdb
    ports:
      - "5432:5432"
    volumes:
      - postgres_data:/var/lib/postgresql/data
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U user -d dappdb"]
      interval: 10s
      timeout: 5s
      retries: 5

  backend:
    build: .
    image: node:16
    container_name: dapp-node
    networks:
      - dapp-for-business_dapp-network
    depends_on:
      db:
        condition: service_healthy
    environment:
      - NODE_ENV=development
      - DATABASE_URL=postgres://user:password@db:5432/dappdb
    ports:
      - "5000:5000"
    volumes:
      - .:/app
    working_dir: /app
    command: "node server.js"

  frontend:
    build:
      context: ./frontend
      dockerfile: Dockerfile
    image: node:16
    container_name: dapp-frontend
    networks:
      - dapp-for-business_dapp-network
    ports:
      - "3000:3000"
    volumes:
      - ./frontend:/app
      - /app/node_modules
    working_dir: /app
    command: "npm start"

volumes:
  postgres_data:
  grafana-data:

networks:
  dapp-for-business_dapp-network:
    external: true
